diff --git a/vllm/v1/worker/gpu/cudagraph_utils.py b/vllm/v1/worker/gpu/cudagraph_utils.py
index e76fe0247f61..c7a7ffe442d2 100644
--- a/vllm/v1/worker/gpu/cudagraph_utils.py
+++ b/vllm/v1/worker/gpu/cudagraph_utils.py
@@ -358,7 +358,9 @@ def forward_fn(cg_mode: CUDAGraphMode) -> None:
                 batch_descriptor = None
                 if cg_mode == CUDAGraphMode.PIECEWISE:
                     assert attn_metadata is None
-                    batch_descriptor = BatchDescriptor(num_tokens=num_tokens)
+                    batch_descriptor = BatchDescriptor(
+                        num_tokens=num_tokens, has_lora=has_lora
+                    )
                 with set_forward_context(
                     attn_metadata,
                     self.vllm_config,
