diff --git a/tensorrt_llm/_torch/model_config.py b/tensorrt_llm/_torch/model_config.py
index 2ceab998b39e..308f079ea1e0 100644
--- a/tensorrt_llm/_torch/model_config.py
+++ b/tensorrt_llm/_torch/model_config.py
@@ -794,7 +794,6 @@ def update_sparse_attention_indexer_config(pretrained_config, kwargs):
                     if checkpoint_window_size is None:
                         checkpoint_window_size = getattr(
                             pretrained_config, 'sliding_window', None)
-                    indexer_k_dtype = sparse_attention_config.indexer_k_dtype if sparse_attention_config else 'fp8'
                     if sparse_attention_config:
                         compress_ratios = sparse_attention_config.compress_ratios
                         window_size = sparse_attention_config.window_size
@@ -838,7 +837,6 @@ def update_sparse_attention_indexer_config(pretrained_config, kwargs):
                         'sparse_attention_config'] = DeepSeekV4SparseAttentionConfig(
                             compress_ratios=compress_ratios,
                             window_size=window_size,
-                            indexer_k_dtype=indexer_k_dtype,
                             **indexer_config)
             else:
                 raise ValueError(
