diff --git a/python/sglang/srt/layers/moe/fused_moe_triton/layer.py b/python/sglang/srt/layers/moe/fused_moe_triton/layer.py
index 943ac4f173bd..9b84319506c2 100644
--- a/python/sglang/srt/layers/moe/fused_moe_triton/layer.py
+++ b/python/sglang/srt/layers/moe/fused_moe_triton/layer.py
@@ -312,9 +312,10 @@ def __init__(
             get_moe_runner_backend().is_flashinfer_trtllm_routed()
             or get_moe_runner_backend().is_flashinfer_trtllm()
         ):
-            logging.warning(
-                "Setting inplace to False for FlashInfer TRTLLM MoE backend."
-            )
+            if self.moe_runner_config.inplace:
+                logging.info(
+                    "Setting inplace to False for FlashInfer TRTLLM MoE backend."
+                )
             self.moe_runner_config.inplace = False
 
         self.should_fuse_routed_scaling_factor_in_topk = (
